lib/mastermind.rb

```diff
@@ -1,40 +1,122 @@
+    @past_guesses = []
+    @restart = false
+  end
+
+  def print_instructions(long_colors)
     puts "\tWelcome to Mastermind! The object of the game is to guess a 4-color code
-        within 10 guesses. The available colors are red, orange, yellow, green,
-        blue, and purple. Colors may be used in the code more than once or not at all.
+        (order matters) within 10 guesses. The available colors are #{long_colors}.
+        Colors may be used in the code more than once or not at all.
 
         To make a guess: type the first letter of each color you're guessing in caps.
-        For example, if you want to guess 'red, orange, yellow, red,' type 'ROYR'.
+        For example, if you want to guess red, orange, yellow, red, enter 'ROYR'.
 
         After each guess, you will receive feedback in the form of colored pins,
         letting you know:
           \t- how many colors are correct but in the wrong position (white pins)
           \t- how many colors are correct and in the right position (red pins)
 
-        To quit the game, type '/q'. To start a new game, type '/r'."
+        To quit the game, type 'quit'. To start a new game, type 'restart'."
+  end
+
+  def loop_guesses(short_colors)
+    feedback = ""
+    while @remaining_guesses > 0
+      puts "You have #{@remaining_guesses} guesses remaining."
+      guess = @game_input.guess(short_colors)
+      quit if guess == "QUIT"
+      if guess == "RESTART"
+        restart_game
+        break
+      end
+
+      feedback = compare(guess)
+      if feedback[:red] == 4
+        win
+        break
+      end
+      log_guess(guess)
+      @remaining_guesses -= 1
+    end
+    loss if feedback[:red] != 4 && @remaining_guesses == 0
+  end
+
+  def log_guess(guess)
+    @past_guesses << guess
+    puts "Your previous guesses were #{@past_guesses}."
   end
 
   def color_count(code)
     color_count = Hash.new(0)
-    code.each do |color|
-      color_key = color.to_sym
+    code.each_char do |c|
+      color_key = c.to_sym
       color_count[color_key] += 1
       end
     color_count
   end
 
-  def get_color_counts(code, guess)
-    code_count = color_count(code)
-    guess = guess.chars
-    guess_count = color_count(guess)
+  def white(code, guess)
+    code_color_count = color_count(code)
+    guess_color_count = color_count(guess)
+
+    white_count = 0
+    guess_color_count.each do |k, v|
+      if v <= code_color_count[k]
+        white_count += v
+      elsif v > code_color_count[k]
+        white_count += code_color_count[k]
+      end
+    end
+    white_count
+  end
+
+  def red_and_white(white_count, code, guess)
+    red_count = 0
+
+    guess.length.times do |i|
+      if guess[i] == code[i]
+        red_count += 1
+        if white_count > 0
+          white_count -= 1
+        end
+      end
+    end
+    puts "Red pins: #{red_count} \nWhite pins: #{white_count}"
+    {:red => red_count, :white => white_count}
+  end
+
+  def compare(guess)
```

Comment:
I think this name here is a bit ambiguous. Think about how someone using your `Mastermind` class would call this:
```rb
mastermind = Mastermind.new(..)
mastermind.compare(...)
```
We're not really sure what we're comparing because there is not much context. If you had this method in a tiny, but focused class called `GuessChecker` or `ScoreKeeper`, then it would make a bit more sense with the extra context. Not a big deal, but something to think about when naming stuff in the future.